proposals/0000-warn-noncanonical-monad-instance-by-default.md

```diff
@@ -0,0 +1,78 @@
+---
+author: Fumiaki Kinoshita
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/314).
+
+# Enable `-Wnoncanonical-monad-instances` and `-Wnoncanonical-monoid-instances` by default
+
+Currently, `-Wnoncanonical-monad-instances` and `-Wnoncanonical-monoid-instances` are not enabled in any of the default, `-Wall` and `-Wcompat`.
+I propose enabling the warning by default.
+
+## Motivation
+
+The prerequisite of the Phase 2 of [monad of no return](https://gitlab.haskell.org/ghc/ghc/wikis/proposal/monad-of-no-return) is to wait until "we're confident that the majority of Hackage has reacted to the warning". However, a warning that's disabled unless you specify it individually gives infinitesimally small incentive to the ecosystem, so we can't really expect them to react. In order to avoid catastrophic breakage like what we experienced when `fail` got removed, I think the warning should be enabled by default, not just in `-Wall` or `-Wcompat`.
+
+I also propose enabling `-Wnoncanonical-monoid-instances` for the same reason.
+
+## Proposed Change Specification
+
+* Add `-Wnoncanonical-monad-instances` and `-Wnoncanonical-monoid-instances` to the set of default warnings.
+
+## Examples
+
+On GHC 8.8 (and 8.11.0.20200108), the following code compiles without warnings.
+
+```haskell
+data P a = P
+
+instance Functor P where
+  fmap _ _ = P
+
+instance Applicative P where
+  pure _ = P
+  _ <*> _ = P
+
+instance Monad P where
+  return _ = P
+  _ >>= _ = P
+```
```

Comment:
It would be helpful to include a `Monoid` instance that triggers `-Wnoncanonical-monoid-instances` in this example.